src/psyclone/domain/lfric/kernel_arg_order.py

```diff
@@ -0,0 +1,1063 @@
+                        # the max size of a stencil branch must be passed
+                        # from the Algorithm layer.
+                        self.stencil_2d_max_extent(meta_arg)
+                    else:
+                        # stencil extent is not provided in the
+                        # metadata so must be passed from the Algorithm
+                        # layer.
+                        self.stencil_unknown_extent(meta_arg)
+                    if meta_arg.stencil == "xory1d":
+                        # if "xory1d is specified then the actual
+                        # direction must be passed from the Algorithm layer.
+                        self.stencil_unknown_direction(meta_arg)
+                    # stencil information that is always passed from the
+                    # Algorithm layer.
+                    if meta_arg.stencil == "cross2d":
+                        self.stencil_2d(meta_arg)
+                    else:
+                        self.stencil(meta_arg)
+            elif type(meta_arg) == OperatorArgMetadata:
+                self.operator(meta_arg)
+            elif type(meta_arg) == ColumnwiseOperatorArgMetadata:
+                self.cma_operator(meta_arg)
+            elif type(meta_arg) == ScalarArgMetadata:
+                self.scalar(meta_arg)
+            else:
+                raise InternalError(
+                    f"Unexpected meta_arg type '{type(meta_arg).__name__}' "
+                    f"found.")
+
+        # For each unique function space (in the order they appear in the
+        # metadata arguments)
+        function_space_args = metadata.meta_args_get(
+            [FieldArgMetadata, FieldVectorArgMetadata,
+             InterGridArgMetadata, InterGridVectorArgMetadata,
+             OperatorArgMetadata, ColumnwiseOperatorArgMetadata])
+        unique_function_spaces = []
+        for arg in function_space_args:
+            if type(arg) in [
+                    OperatorArgMetadata, ColumnwiseOperatorArgMetadata]:
+                if arg.function_space_to not in unique_function_spaces:
+                    unique_function_spaces.append(arg.function_space_to)
+                if arg.function_space_from not in unique_function_spaces:
+                    unique_function_spaces.append(arg.function_space_from)
+            else:
+                if arg.function_space not in unique_function_spaces:
+                    unique_function_spaces.append(arg.function_space)
+
+        for function_space in unique_function_spaces:
+            # Provide arguments common to LMA operators and fields on
+            # a space *unless* this is an inter-grid or CMA
+            # matrix-matrix kernel
+            if metadata.kernel_type not in [
+                    "cma-matrix-matrix", "inter-grid"]:
+                self.fs_common(function_space)
+
+            # Provide additional arguments if there is a
+            # field on this space
+            if [arg for arg in metadata.meta_args_get(
```

Comment:
I'm assuming there is no FunctionSpace class as that is associated with the old metadata.

Reply:
Perhaps things will need to change when we start trying to replace the existing lfric metadata implementation.

Reply:
Whilst nothing will change on this in this PR there is perhaps a distinction that could be usefully made. I think a class representing a `field` is different to a class representing field metadata. I do agree that it would be nicer to store metadata as enumerated types rather than strings and that should be done soon before the new metadata starts being used too much.

Reply:
OK. I shall try to remember to keep an eye on this.